### ISSUE TITLE
Multiprocessed randomized_search and grid_search?

### ISSUE THREAD

**Issue by kmedved** [catboost/catboost#1493]
Hello - thanks for all the work on this amazing state-of-the-art package.

I am interested in whether there is any ability or any plans to implement multiprocessed hyperparameter tuning via Catboost's `randomized_search` or `grid_search` functionality? Presently, it looks like these will only search one set of parameters at a time, but there are obviously potential speed gains to be had from searching multiple sets of parameters simultaneously. 

Generally, given the sequential nature of GBDTs generally, there seems to be extremely diminishing returns to using more cores per single model fit, so if tuning on a 16-core machine, it would be more efficient to run 16 trials with a `thread_count` parameter of 1 than 1 trial with a `thread_count` parameter of 16. (Please let me know if I'm mistaken about this, but that's consistent with my testing).

Sklearn's `randomized_search` and `grid_search` are multiprocessed via joblib, which works well, but does not support passing a validation set for the purposes of early stopping, which is obviously key to many GBDTs including Catboost. As result, hyperparameter searching via those tools is lacking as well. Other hyperparamter tuning libraries such as Optuna or HyperOpt only seem to support multiprocessing via a cluster, and are difficult to locally multiprocess. ([Issue 1387](https://github.com/catboost/catboost/issues/1387) is somewhat related to this as well.)
